### PR TITLE
EUI-5888

### PR DESCRIPTION
**Pull request by SivaprakasamSivaprakasam**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/EUI-5888

### Change description ###
Link cases- Tech ticket on linked cases tab (reasons code)

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/shared/components/palette/case-link/domain/linked-cases.model.ts [hunk 1]
```diff
@@ -32,3 +32,20 @@ export class LinkReason {
   reason: string;
   otherDescription?: string;
 }
+
+export class GetLinkedCases {
```
Comment (johnbenjamin-hmcts) [inline src/shared/components/palette/case-link/domain/linked-cases.model.ts:36]: The naming convention of the class name `GetLinkedCases` looks like a property or method name.

File: src/shared/components/palette/case-link/domain/linked-cases.model.ts [hunk 2]
```diff
@@ -32,3 +32,20 @@ export class LinkReason {
...
+  linkedCases: []
+}
+export class GetLinkedCasesCaseLinkResponse {
```
Comment (johnbenjamin-hmcts) [inline src/shared/components/palette/case-link/domain/linked-cases.model.ts:39]: The naming convention of the class name `GetLinkedCasesCaseLinkResponse` looks like a property or method name.

File: src/shared/components/palette/case-link/domain/linked-cases.model.ts (hunk 1 above)
Comment (johnbenjamin-hmcts) [reply src/shared/components/palette/case-link/domain/linked-cases.model.ts:36]: The class properties missing semi-colon.

File: src/shared/components/palette/case-link/domain/linked-cases.model.ts (hunk 2 above)
Comment (johnbenjamin-hmcts) [reply src/shared/components/palette/case-link/domain/linked-cases.model.ts:39]: The class properties missing semi-colon.

File: src/shared/components/palette/case-link/domain/linked-cases.model.ts [hunk 3]
```diff
@@ -32,3 +32,20 @@ export class LinkReason {
...
+  caseNameHmctsInternal: string
+  caseReference: string
+  ccdCaseTyp: string
+  ccdJurisdiction: string
+  state: string
+  linkDetails: [LinkDetails]
+}
+
+export class LinkDetails {
+    createdDateTime: Date
```
Comment (johnbenjamin-hmcts) [inline src/shared/components/palette/case-link/domain/linked-cases.model.ts:49]: The class properties missing semi-colon.

**Review by johnbenjamin-hmcts — CHANGES_REQUESTED** [review]
Minor changes requested

**Review by johnbenjamin-hmcts — APPROVED** [review]
Approved

File: src/shared/components/palette/case-link/components/linked-cases-table/linked-cases-to-table.component.ts [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+  public getCaseRefereneLink(caseRef) {
+    return caseRef.slice(this.caseId.length - 4);
+  }
+
+  sortByReasonCode() {
+    const topLevelresultArray = [];
+    let secondLevelresultArray = [];
+    const data = this.caseField.value || [];
+    data.forEach((item: any) => {
```
Comment (rayliangatos) [inline src/shared/components/palette/case-link/components/linked-cases-table/linked-cases-to-table.component.ts:74]: could this any to be a type?

File: src/shared/components/palette/case-link/components/linked-cases-table/linked-cases-to-table.component.ts [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+          searchCases.forEach(response => {
+          response.results.forEach((result: any) =>
+            this.linkedCasesFromResponse.push(this.mapResponse(result)));
+        });
+        this.isLoaded = true;
+      });
+    }
+  }
+
+  public searchCasesByCaseIds(searchCasesResponse: any[]) {
```
Comment (rayliangatos) [inline src/shared/components/palette/case-link/components/linked-cases-table/linked-cases-to-table.component.ts:108]: return type?

File: src/shared/components/palette/case-link/components/linked-cases-table/linked-cases-to-table.component.ts [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    return forkJoin(searchCasesResponse);
+  }
+
+  hasLeadCaseOrConsolidated(reasonCode: string) {
```
Comment (rayliangatos) [inline src/shared/components/palette/case-link/components/linked-cases-table/linked-cases-to-table.component.ts:112]: modifier and return type

File: src/shared/components/palette/case-link/components/linked-cases-table/linked-cases-to-table.component.ts [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    return reasonCode === 'Progressed as part of this lead case' || reasonCode === 'Case consolidated'
+  }
+
+  mapResponse(esSearchCasesResponse) {
```
Comment (rayliangatos) [inline src/shared/components/palette/case-link/components/linked-cases-table/linked-cases-to-table.component.ts:116]: modifier and return type

File: src/shared/components/palette/case-link/components/linked-cases-table/linked-cases-to-table.component.ts [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    })
+    return topLevelresultArray.concat(secondLevelresultArray)
+  }
+
+  public getAllLinkedCaseInformation() {
+    const searchCasesResponse = [];
+    const sortedCases = this.sortByReasonCode()
+    const linkedCaseIds = this.groupByCaseType(sortedCases, 'caseType');
+    Object.keys(linkedCaseIds).forEach(key => {
+      const esQuery = this.constructElasticSearchQuery(linkedCaseIds[key], 100)
```
Comment (rayliangatos) [inline src/shared/components/palette/case-link/components/linked-cases-table/linked-cases-to-table.component.ts:93]: is hardcoded to size 100 what is required?

File: src/shared/components/palette/case-link/components/linked-cases-table/linked-cases-to-table.component.ts [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    const caseReasonCode = this.caseField.value.find(item => item.caseReference === esSearchCasesResponse.case_id);
+    caseReasonCode.reasons.forEach(code => {
+      const foundReasonMapping = this.linkedCaseReasons.find(reason => reason.key === code.reasonCode);
+      if (foundReasonMapping) {
+        reasonDescriptons.push(foundReasonMapping.value_en);
+      }
+    });
+    return {
+      case_id: esSearchCasesResponse.case_id,
+      caseName: '',
```
Comment (rayliangatos) [inline src/shared/components/palette/case-link/components/linked-cases-table/linked-cases-to-table.component.ts:127]: caseName can be removed as it's not existing

File: src/shared/components/palette/case-link/components/linked-cases-table/linked-cases-to-table.component.ts [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+      caseType: esSearchCasesResponse.case_fields['[CASE_TYPE]'],
+      service: esSearchCasesResponse.case_fields['[JURISDICTION]'],
+      state: esSearchCasesResponse.case_fields['[STATE]'],
+      reasons: reasonDescriptons,
+    } as unknown as LinkedCasesResponse
+  }
+
+  constructElasticSearchQuery(caseIds: any[], size: number) {
```
Comment (rayliangatos) [inline src/shared/components/palette/case-link/components/linked-cases-table/linked-cases-to-table.component.ts:135]: modifier and return type

File: src/shared/components/palette/case-link/domain/linked-cases.model.ts (hunk 1 above)
Comment (rayliangatos) [reply src/shared/components/palette/case-link/domain/linked-cases.model.ts:36]: agreed with John's comments

File: src/shared/services/search/search.service.ts [hunk 11]
```diff
@@ -41,6 +41,28 @@ export class SearchService {
       );
   }
 
+  public searchCasesByIds(
```
Comment (rayliangatos) [inline src/shared/services/search/search.service.ts:44]: return type?

**Review by rayliangatos — COMMENTED** [review]
The whole pr I couldn't see it call the LoV api to transform the code to value, anything misunderstanding?

**Review by rayliangatos — CHANGES_REQUESTED** [review]
couldn't see it call the LoV api to get list of value for reasons.

**Review by rayliangatos — APPROVED** [review]
Approved with some remaining unresolved comments

**Pull request merged**